### PR TITLE
Allow CBMs to specify mutations that prevent installation.

### LINKED ISSUE #46013: Allow mutations to check for CBMs (if they don't already do so)
# Is your feature request related to a problem? Please describe.
I'm not entirely certain how difficult it would be to allow mutations to check for CBMs on addition, as I was told it was currently impossible, but CBMs can already do this the other way. Also, I don't know how needed this is, but I thought it would be interesting to add if a moxpore mutation line gets added, so that the mutations can adapt to the CBMs currently installed (as moxpores are machines, I figured it would be cool if they interfaced with other machines).

# Describe the solution you'd like
Allow mutations to check for CBMs on addition, similar to how CBMs can check for mutations on installation.

# Describe alternatives you've considered
Not doing this, probably because it would be too hard.

# Additional context
NA

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
Infrastructure "Allow CBMs to specify mutations that prevent installation"

#### Purpose of change
Resolves #46013
https://github.com/CleverRaven/Cataclysm-DDA/pull/47117#issuecomment-770947180

#### Describe the solution
First commit: Move code around to make this slightly cleaner.
Second commit: Add a set of conflicting mutations to bionics, add it to bionic rejection reasons, prevent mutating it.

See commits for more detail.

#### Testing
Try installing CBMs on the autodoc, with various reasons for them to be rejected. See that they are appropriately rejected.
Make a CBM require `HUGE`, mutate it, and see you cannot install the CBM.
Remove the huge mutation, install the CBM, and then chug ursine serum. See that you never mutate `Freakishly Huge`.

(I did the above with the Alarm System CBM here).

image

### CODE REVIEW

**Comment by Fris0uman** [comment]
Maybe this should use something more generic, like first look if mutation and CBM affect the same bodyparts, then look if the CBM or the mutation has a EXCLUDE_MUTATION/CBM flag and then do the exclusion? Because here filling the fields for every CBMs might get tedious and some CBM might get through the cracks.

**Comment by Qrox** [comment]
> Remove the huge mutation, install the CBM, and then chug ursine serum. See that you never mutate Freakishly Huge.

I don't think the alarm system CBM is capable of keeping you from mutating into a huge bear. It would make more sense for a mutation to render conflicting CBMs disabled or malfunctional (similar to how failed installation attempts add malfunctional bionics).

**Comment by anothersimulacrum** [comment]
> > Remove the huge mutation, install the CBM, and then chug ursine serum. See that you never mutate Freakishly Huge.
> 
> I don't think the alarm system CBM is capable of keeping you from mutating into a huge bear. It would make more sense for a mutation to render conflicting CBMs disabled or malfunctional (similar to how failed installation attempts add malfunctional bionics).

Neither do I, it's just the example I used for testing, more invasive bionics would probably be the ones that are problematic.

Part of me would prefer to wait until mutation takes time and gives more signals to do more than just block mutation, because if I add consequences, they're not going to be light, and it's something that can happen with basically no warning.

I'm not unwilling to do it here, but I do want to be careful with it.

File: src/bionics.cpp [hunk 1]
```diff
@@ -2317,6 +2319,54 @@ bool Character::uninstall_bionic( const bionic &target_cbm, monster &installer,
...
+
+    if( it->has_flag( flag_FILTHY ) ) {
+        // NOLINTNEXTLINE(cata-text-style): single space after the period for symmetry
+        const std::string msg = by_autodoc ? _( "/!\\ CBM is highly contaminated. /!\\" ) :
+                                _( "CBM is filthy." );
+        return ret_val<bool>::make_failure( msg );
+    } else if( it->has_flag( flag_NO_STERILE ) ) {
+        // NOLINTNEXTLINE(cata-text-style): single space after the period for symmetry
+        const std::string msg = by_autodoc ?
+                                _( "/!\\ CBM is not sterile. /!\\ Please use autoclave to sterilize." ) :
```
Comment (ZhilkinSerg) [inline src/bionics.cpp:2340]:
```suggestion
                                _( "/!\\ CBM is not sterile.  /!\\ Please use autoclave to sterilize." ) :
```

Comment (Qrox) [inline src/bionics.cpp:2340]:
```suggestion
        const std::string msg = by_autodoc ?
                                // NOLINTNEXTLINE(cata-text-style): single space after the period for symmetry
                                _( "/!\\ CBM is not sterile. /!\\ Please use autoclave to sterilize." ) :
```

Comment (Qrox) [reply src/bionics.cpp:2340]: The string is correct, but the `nolint` comment is not at the correct place.

**Pull request merged**